mephisto/core/task_launcher.py

```diff
@@ -46,74 +53,95 @@ def __init__(
-            yield data
-
-    def _try_create_assignments(self) -> None:
-        """ Create an assignment and associated units for the generated assignment data """
-        for data in self.generate_assignments():
-            task_run = self.task_run
-            task_config = task_run.get_task_config()
-            assignment_id = self.db.new_assignment(
+        assignment = Assignment(self.db, assignment_id)
+        assignment.write_assignment_data(assignment_data)
+        self.assignments.append(assignment)
+        unit_count = len(assignment_data["unit_data"])
+        for unit_idx in range(unit_count):
+            unit_id = self.db.new_unit(
                 task_run.task_id,
                 task_run.db_id,
                 task_run.requester_id,
-                task_run.task_type,
+                assignment_id,
+                unit_idx,
+                task_config.task_reward,
                 task_run.provider_type,
+                task_run.task_type,
                 task_run.sandbox,
             )
-            assignment = Assignment(self.db, assignment_id)
-            assignment.write_assignment_data(data)
-            self.assignments.append(assignment)
-            unit_count = len(data["unit_data"])
-            for unit_idx in range(unit_count):
-                unit_id = self.db.new_unit(
-                    task_run.task_id,
-                    task_run.db_id,
-                    task_run.requester_id,
-                    assignment_id,
-                    unit_idx,
-                    task_config.task_reward,
-                    task_run.provider_type,
-                    task_run.task_type,
-                    task_run.sandbox,
-                )
-                self.units.append(Unit(self.db, unit_id))
+            self.units.append(Unit(self.db, unit_id))
+            with self.unlaunched_units_access_condition:
                 self.unlaunched_units[unit_id] = Unit(self.db, unit_id)
-                self.keep_launching_units = True
-            time.sleep(ASSIGNMENT_GENERATOR_WAIT_SECONDS)
-        self.assignment_exist = False
+            self.keep_launching_units = True
+
+    def _try_generate_assignments(self) -> None:
+        """ Try to generate more assignments from the assignments_data_iterator"""
+        try:
+            data = next(self.assignment_data_iterable)
+            self._create_single_assignment(data)
+        except StopIteration:
+            self.did_create_assignments = False
+            self.did_launch_units = False
+            self.finished_generators = True
+
+    def create_assignments(self) -> None:
+        """ Create an assignment and associated units for the generated assignment data """
+        if self.generator_type == GeneratorType.none:
+            for data in self.assignment_data_iterable:
+                self._create_single_assignment(data)
+        else:
+            self.did_create_assignments = True
+            self.did_launch_units = True
```

Comment:
You can actually have the `assignments` thread be launched in the same way as the `units` thread at this point, as both need to be run separately anyways.